libbeat/common/mapstr.go

```diff
@@ -134,3 +152,61 @@ func cleanUpMapValue(v interface{}) interface{} {
 		return fmt.Sprintf("%v", v)
 	}
 }
+
+// MergeFields merges the top-level keys and values in each source hash (it does
+// not perform a deep merge). If the same key exists in both, the value in
+// fields takes precedence. If underRoot is true then the contents of the fields
+// MapStr is merged with the value of the 'fields' key in ms.
+//
+// An error is returned if underRoot is true and the value of ms.fields is not a
+// MapStr.
+func MergeFields(ms, fields MapStr, underRoot bool) error {
+	if ms == nil || fields == nil {
+		return nil
+	}
+
+	fieldsMS := ms
+	if !underRoot {
+		f, ok := ms[FieldsKey]
+		if !ok {
+			fieldsMS = make(MapStr, len(fields))
+			ms[FieldsKey] = fieldsMS
+		} else {
+			// Use existing 'fields' value.
+			fieldsMS, ok = f.(MapStr)
+			if !ok {
+				return ErrorFieldsIsNotMapStr
+			}
+		}
+	}
+
+	// Add fields and override.
```

Comment:
Filebeat logged a warning when overwriting. I think warning would be too verbose if overwriting were done on purpose (either to overwrite a field defined in the shipper config or to overwrite one of the events other fields). The debug level would be ok if it were isolated to its own selector. But I'm thinking to omit it to keep the publishing path as fast a possible. If you feel strongly about it, I'm happy to add it.